### DISCUSSION TITLE
[PartialTransformer] Field Option[Field].None to Proto3.Fields.Default

### DISCUSSION THREAD

**Discussion by ps-rterzman** [scalalandio/chimney#474] [Q&A] [answered]
Hello, 
I am trying to use PartialTransfomer for achieving domain to proto3 transfomration.
I faced an issue related to missed fields in domain case class even if proto3 fields are optional with default value.

**Domain case class**
`final case class ClientProductEntity(id: Option[String], promotions: Option[Seq(Promo)], PSCategory: Option[String])
final case class Promo(effectiveStartDate: Option[Long], effectiveEndDate: Option[Long], map: Seq[Map])`

**Proto3**
`@SerialVersionUID(0L)
final case class ClientProduct( id: _root_.scala.Predef.String = "", releaseDate: _root_.scala.Option[_root_.scala.Long] = _root_.scala.None, promotions: _root_.scala.Seq[com.pricespider.unity.domain.ClientProduct.Promo] = _root_.scala.Seq.empty), pSCategory: _root_.scala.Option[_root_.scala.Predef.String] = _root_.scala.None)`

PartialTransformers:
`    val mappingResult: Either[Result.Errors, ClientProduct] = clientProductEntity
      .intoPartial[ClientProduct]
      .withFieldRenamed(_.PSCBID, _.pSCBID)
      .withFieldRenamed(_.PSCID, _.pSCID)
      .withFieldRenamed(_.PSCPID, _.pSCPID)
      .withFieldRenamed(_.PSCategory, _.pSCategory)
      .withFieldComputed(_.pSPID, from => from.PSPID.orNull)
      .withFieldConst(_.unknownFields, _root_.scalapb.UnknownFieldSet.empty)
      .enableOptionDefaultsToNone
      .transform
      .asEither
      .asOption
      .orNull`
     
mappingResult.right is None 
mappingResult.left contains all missed values

I am wondering is there any way to map missed | Option values with default values. Could you help me with that?

**Comment by MateuszKubuszok** [comment, marked as answer]
With the information code I cannot reproduce anything as I am missing types (`Seq[Map]`, `com.pricespider.unity.domain.ClientProduct.Promo`), fields (`ClientProductEntity` has no `PSCBID`, `PSCID`) and in general this code doesn't compile (`Option[Seq(Promo)]` instead of `Option[Seq[Promo]]`). So I can only guess, what you want to achieve.

From the description, an I can only guess that you wanted to use default values although [default values are used](https://chimney.readthedocs.io/en/stable/supported-transformations/#allowing-fallback-to-the-constructors-default-values) when there is no source field, not when the source field is `None`. There is no support for using default values as `None` handling, however for this particular case:

```scala
implicit val noneIntIsZero: Transformer[Option[Int], Int] = _.getOrElse(0)
implicit val noneLongIsZero: Transformer[Option[Long], Long] = _.getOrElse(0L)
implicit val noneStringIsEmpty: Transformer[Option[String], String] = _.getOrElse("")

clientProductEntity.into[ClientProduct]....transform
```

would handle most cases.

I don't see how `PartialTransformer` would be useful here since transformation from a domain model to a Protobuf cannot fail.

**Reply by ps-rterzman** [reply]
Thank you for your answer. 
I tried to use a Transformer instead. Here what I have:
`    
// Convert Option[Seq[A]] to Seq[B]
implicit def optionSeqAToSeqB[A, B](implicit aToB: Transformer[A, B]): Transformer[Option[Seq[A]], Seq[B]] =
      (from) => from.getOrElse(Seq.empty).map(aToB.transform)
    implicit val optToString:         Transformer[Option[String], String]                             = from => from.getOrElse("")
    implicit val optToLong:           Transformer[Option[Long], Long]                                 = from => from.getOrElse(0L)
    implicit val optToInt:            Transformer[Option[Int], Int]                                   = from => from.getOrElse(0)`
    
 I **get an error** during compilation time: 
 `com.transformer.domain.ClientProduct
  map: scala.collection.immutable.Seq[com.transformer.domain.ClientProduct.Map] - can't derive transformation from map: scala.Option[scala.collection.immutable.Seq[com.transformer.model.Map]] in source type com.pricespider.unity.model.ClientProductEntity
  promotions: scala.collection.immutable.Seq[com.transformer.domain.ClientProduct.Promo] - can't derive transformation from promotions: scala.Option[scala.collection.immutable.Seq[com.transformer.model.Promo]] in source type com.pricespider.unity.model.ClientProductEntity`
 
package com.transformer.domain:
`
    final case class ClientProductEntity(PSCID:  Option[String],
    PSCPID:  Option[String],
    PSPID:  Option[String],
    PSCBID:  Option[String],
    name:  Option[String],
    map:  Option[Seq[Map]],
    promotions:  Option[Seq[Promo]],
    releaseDate: Option[_root_.scala.Long],
    createdBy: Option[AuditBy],
    revision: Option[Int])
   
   final case class Map(amount: Option[Double], currencyCode: Option[String])
   final case class Promo(effectiveStartDate: Option[Long], effectiveEndDate: Option[Long], map: Seq[Map])
   final case class AuditBy(source: Option[String], userId: Option[String], traceID: Option[String], spanID: Option[String])`

package com.transformer.model:
`
    syntax = "proto3";
    option java_package = "com.transformer.domain";

    message ClientProduct {
    string PSCPID = 1;
    string PSPID = 2;
    string PSCBID = 3;
    string name = 4;
    repeated .ClientProduct.Map map = 5;
    repeated .ClientProduct.Promo promotions = 6;
    optional int64 releaseDate = 7;
    .ClientProduct.By createdBy = 8;
    int32 revision = 9;
    }

    message Map {
    double amount = 1;
    string currencyCode = 2;
    }

    message Promo {
    int64 effectiveStartDate = 1;
    optional int64 effectiveEndDate = 2;
    repeated .ClientProduct.Map map = 3;
    }

    message By {
    string source = 1;
    string userId = 2;
    string traceID = 3;
    string spanID = 4;
    }

    @SerialVersionUID(0L) 
    final case class ClientProduct(
    pSCPID: _root_.scala.Predef.String = "",
    pSPID: _root_.scala.Predef.String = "",
    pSCBID: _root_.scala.Predef.String = "",
    name: _root_.scala.Predef.String = "",
    map: _root_.scala.Seq[com.pricespider.unity.domain.ClientProduct.Map] = _root_.scala.Seq.empty,
    promotions: _root_.scala.Seq[com.pricespider.unity.domain.ClientProduct.Promo] = _root_.scala.Seq.empty,
    releaseDate: _root_.scala.Option[_root_.scala.Long] = _root_.scala.None,
    createdBy: _root_.scala.Option[com.pricespider.unity.domain.ClientProduct.By] = _root_.scala.None,
    revision: _root_.scala.Int = 0)
    
      @SerialVersionUID(0L) 
      final case class Map(
      amount: _root_.scala.Double = 0.0,
      currencyCode: _root_.scala.Predef.String = "",
      unknownFields: _root_.scalapb.UnknownFieldSet = _root_.scalapb.UnknownFieldSet.empty
      ) extends scalapb.GeneratedMessage
    
      
      @SerialVersionUID(0L) 
      final case class Promo(
      effectiveStartDate: _root_.scala.Long = 0L,
      effectiveEndDate: _root_.scala.Option[_root_.scala.Long] = _root_.scala.None,
      map: _root_.scala.Seq[com.pricespider.unity.domain.ClientProduct.Map] = _root_.scala.Seq.empty,
      unknownFields: _root_.scalapb.UnknownFieldSet = _root_.scalapb.UnknownFieldSet.empty
      ) extends scalapb.GeneratedMessage
      
      @SerialVersionUID(0L) 
      final case class By(
      source: _root_.scala.Predef.String = "",
      userId: _root_.scala.Predef.String = "",
      traceID: _root_.scala.Predef.String = "",
      spanID: _root_.scala.Predef.String = "",
      unknownFields: _root_.scalapb.UnknownFieldSet = _root_.scalapb.UnknownFieldSet.empty
      ) extends scalapb.GeneratedMessage
      
`
  P.S. I am really sorry about code formatting. I tried to do my best.

**Reply by MateuszKubuszok** [reply]
Sorry for the late reply.

You error is:

```
com.transformer.domain.ClientProduct
  map: scala.collection.immutable.Seq[com.transformer.domain.ClientProduct.Map] - can't derive transformation from map: scala.Option[scala.collection.immutable.Seq[com.transformer.model.Map]] in source type com.pricespider.unity.model.ClientProductEntity
  promotions: scala.collection.immutable.Seq[com.transformer.domain.ClientProduct.Promo] - can't derive transformation from promotions: scala.Option[scala.collection.immutable.Seq[com.transformer.model.Promo]] in source type com.pricespider.unity.model.ClientProductEntity
```

The issue is here:

```scala
implicit def optionSeqAToSeqB[A, B](implicit aToB: Transformer[A, B]): Transformer[Option[Seq[A]], Seq[B]] =
  (from) => from.getOrElse(Seq.empty).map(aToB.transform)
```

due to specifics of [how derivation in Chimney works](https://chimney.readthedocs.io/en/stable/under-the-hood/?h=autoderived#dsl) your code does not call a macro recursively. You'd have to modify your code to:

```scala
implicit def optionSeqAToSeqB[A, B](implicit aToB: Transformer.AutoDerived[A, B]): Transformer[Option[Seq[A]], Seq[B]] =
  (from) => from.getOrElse(Seq.empty).map(aToB.transform)
```

then macro could derive `Transformer.AutoDerived[A, B]`, this implicit would find it, and macro would find this implicit.
